tools/docker.go

```diff
@@ -38,18 +40,34 @@ func DockerRun(ctx context.Context, args []string, volumes []string) error {
 	}
 
 	// check docker installation
-	// is docker installed?
-	//
+	if err := exec.CommandContext(ctx, dockerPath).Run(); err != nil {
+		return fmt.Errorf("could not run docker: %v", err)
+	}
+
 	// are we running outside docker?
-	//
+	if FileExists("/.dockerenv") == nil {
+		return fmt.Errorf("running inside docker. Set VSYNCER_DOCKER=false")
+	}
+
 	// is it rootless?
-	// docker info -f "{{println .SecurityOptions}}" | grep rootless
-	//
+	if output, err := exec.CommandContext(ctx, dockerPath, "info", "-f",
+		"{{println .SecurityOptions}}").Output(); err != nil {
+		return fmt.Errorf("could not run docker: %v", err)
+	} else {
+		rootless = strings.Contains(string(output), "rootless")
+	}
+
 	// if not rooless do I have permission?
-	// check if user in docker group, otherwise should we request sudo?
+	if !rootless {
+		// check if user in docker group, otherwise should we request sudo?
+		if output, err := exec.CommandContext(ctx, "id", "-Gn").Output(); err != nil {
+			return fmt.Errorf("could get user groups: %v", err)
+		} else if !strings.Contains(string(output), "docker") {
+			return fmt.Errorf("user is not in docker group")
+		}
 
-	// decide whether we should set uid:gid
-	cmd = append(cmd, "-u", fmt.Sprintf("%v:%v", u.Uid, u.Gid))
+		cmd = append(cmd, "-u", fmt.Sprintf("%v:%v", u.Uid, u.Gid))
```

Comment:
I'm thinking maybe it would even be better if we removed the  `-u$uid:$gid`, and would let vsyncer docker generate the temporary .ll files into a directory which is **not** on the mounted volume. vsyncer only really needs to read files from the mounted directory after all.

But we can also leave this for after your talk.

Reply:
Yes, but as we discussed offline, often we want to actually inspect the intermediate files. So we want to generate them in the mounted volume and for that we need the right user permissions.